lib/core/connect.js

```diff
@@ -76,7 +76,7 @@ function buildConnector ({ maxCachedSessions, socketPath, timeout, ...opts }) {
     }
 
     const timeoutId = timeout
-      ? setTimeout(onConnectTimeout, timeout, socket)
+      ? setTimeout(() => setImmediate(() => onConnectTimeout(socket)), timeout)
```

Comment:
I'm a little unsure about this... but at least we need a comment here explaining why.

Reply:
sure, how about something like 

"setImmediate is needed to make sure that error event callbacks are executed before the TO socket destroy logic."?

Reply:
> I'm a little unsure about this...
I was also not very sure about this but thinking of it again: Deferring throwing a TO in the hope that we already got a response makes sense.